analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt

```diff
@@ -0,0 +1,194 @@
+            repoConfig: RepositoryConfiguration
+        ) = SpdxDocumentFile(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    private fun SpdxPackage.toIdentifier() =
+        Identifier(
+            type = managerName,
+            namespace = originator.extractOrganization().orEmpty(),
+            name = name,
+            version = versionInfo
+        )
+
+    private fun String.extractOrganization() =
+        lineSequence().mapNotNull { line ->
+            line.removePrefix("Organization: ").takeIf { it != line }
+        }.firstOrNull()
+
+    /**
+     * Return the dependencies of [pkg] defined in [doc] of the [SpdxRelationship.Type.DEPENDENCY_OF] type.
+     */
+    private fun getDependencies(pkg: SpdxPackage, doc: SpdxDocument): SortedSet<PackageReference> =
+        getDependencies(pkg, doc, SpdxRelationship.Type.DEPENDENCY_OF) { target ->
+            val dependency = doc.packages.single { it.spdxId == target }
+            PackageReference(
+                id = dependency.toIdentifier(),
+                dependencies = getDependencies(dependency, doc)
+            )
+        }
+
+    /**
+     * Return the dependencies of [pkg] defined in [doc] of the given [dependencyOfRelation] type. Optionally, the
+     * [SpdxRelationship.Type.DEPENDS_ON] type is handled by [dependsOnCase].
+     */
+    private fun getDependencies(
+        pkg: SpdxPackage,
+        doc: SpdxDocument,
+        dependencyOfRelation: SpdxRelationship.Type,
+        dependsOnCase: (String) -> PackageReference? = { null }
+    ): SortedSet<PackageReference> =
+        doc.relationships.mapNotNullTo(sortedSetOf()) { (source, target, relation) ->
+            when {
+                // Dependencies can either be defined on the target...
+                target == pkg.spdxId && relation == dependencyOfRelation -> {
+                    val dependency = doc.packages.single { it.spdxId == source }
+                    PackageReference(
+                        id = dependency.toIdentifier(),
+                        dependencies = getDependencies(dependency, doc, dependencyOfRelation, dependsOnCase)
+                    )
+                }
+
+                // ...or on the source.
+                source == pkg.spdxId && relation == SpdxRelationship.Type.DEPENDS_ON -> {
+                    dependsOnCase(target)
+                }
+
+                else -> null
+            }
+        }
+
+    /**
+     * Return a [Scope] created from the given type of [relation] for [projectPackage] in [spdxDocument], or `null` if
+     * there are no such relations.
+     */
+    private fun createScope(spdxDocument: SpdxDocument, projectPackage: SpdxPackage, relation: SpdxRelationship.Type) =
+        getDependencies(projectPackage, spdxDocument, relation).takeUnless { it.isEmpty() }?.let { dependencies ->
+            Scope(
+                name = relation.name.removeSuffix("_DEPENDENCY_OF").toLowerCase(),
```

Comment:
Imo `removeSuffix("_DEPENDENCY_OF"` is a hidden assumption about the callers.
How about eliminating that by "inlining" the callers by turning the signature into s.th. like `fun createScopes(spdxDocument: SpdxDocument, projectPackage: SpdxPackage): List<Scope>

Reply:
What would `createScopes()` (plural) then do internally? Wouldn't it just iterate over all scopes / relations, which basically is what I do just "one level up" in `resolveDependencies()`?

I think it's ok to make assumptions about the relation name here, it's a private function after all. Also, it's not really a "hard" assumption in the sense that it would fail if the relation name would *not* end with "_DEPENDENCY_OF"; just if it does, the suffix is removed.

Reply:
Agree that it makes sense to keep it as is.